scripts/run_portserver_test.py

```diff
@@ -0,0 +1,454 @@
+        swap_get_port = self.swap(
+            run_portserver.PortPool, 'get_port_for_process',
+            lambda *unused_args: 0)
+        swap_should_allocate_port = self.swap(
+            run_portserver, 'should_allocate_port', lambda _: True)
+        with self.swap_log, swap_get_port, swap_should_allocate_port:
+            request_handler.handle_port_request(1010)
+            request_handler.dump_stats()
+
+        self.assertIn('denied-allocations 1', self.terminal_logs)
+
+    def test_port_server_request_handler_allocates_port_to_client(
+            self) -> None:
+        request_handler = run_portserver.PortServerRequestHandler((8181,)) # type: ignore[no-untyped-call]
+        swap_get_port = self.swap(
+            run_portserver.PortPool, 'get_port_for_process',
+            lambda *unused_args: 8080)
+        swap_should_allocate_port = self.swap(
+            run_portserver, 'should_allocate_port', lambda _: True)
+        with self.swap_log, swap_get_port, swap_should_allocate_port:
+            request_handler.handle_port_request(1010)
+            request_handler.dump_stats()
+
+        self.assertIn('total-allocations 1', self.terminal_logs)
+
+    def test_failure_to_start_server_throws_error(self) -> None:
+        class FailingMockSocket(MockSocket):
+            """Socket that fails while invoking bind()."""
+
+            def bind(self, *unused_args: Any) -> None: # pylint: disable=missing-docstring
+                raise socket.error('Some error occurred.')
+
+        def dummy_handler(data: int) -> str:
+            return str(data)
+
+        swap_socket = self.swap(
+            socket, 'socket', lambda *unused_args: FailingMockSocket())
+        error_msg = (
+            'Failed to bind socket {}. Error: {}'.format(
+                8181, socket.error('Some error occurred.')))
+        with swap_socket, self.assertRaisesRegex(RuntimeError, error_msg):
+            run_portserver.Server(dummy_handler, '8181') # type: ignore[no-untyped-call]
+
+    def test_server_closes_gracefully(self) -> None:
+        mock_socket = MockSocket()
+        mock_socket.port = '\08181'
+
+        def dummy_handler(data: Any) -> str:
+            return str(data)
+        swap_hasattr = self.swap(
+            builtins, 'hasattr', lambda *unused_args: False)
+        swap_socket = self.swap(
+            socket, 'socket', lambda *unused_args: mock_socket)
+
+        with swap_socket, swap_hasattr:
+            server = run_portserver.Server(dummy_handler, '\08181') # type: ignore[no-untyped-call]
+            run_portserver.Server.handle_connection(MockSocket(), dummy_handler)
+            server.close()
+
+        self.assertTrue(server.socket.server_closed)
+
+    def test_server_on_close_removes_the_socket_file(self) -> None:
+        path = '8181'
+        def dummy_handler(data: Any) -> str:
+            return str(data)
+        swap_hasattr = self.swap(
```

Comment:
Use `swap_with_checks` to make sure this gets called as expected. Ditto elsewhere (e.g. for `sys.exit` below)

Reply:
Added `swap_with_checks` here.
For the `sys.exit` below, when we create a mock that does nothing when it hits `sys.exit`, the execution falls down and eventually hits another `sys.exit` statement in the `main` method. Both of them of are called with different args (1 and 0) hence it is not possible to create a `swap_with_checks` that handles both the occurrences.